### PR TITLE
build: use -trimpath in non-DBG mode

### LINKED ISSUE #110928: kubernetes-1.24.0 reproducible builds: kubelet binary varies
### What happened?

While working on [reproducible builds](https://reproducible-builds.org/) for [openSUSE](https://en.opensuse.org/openSUSE:Reproducible_Builds), I found that our kubernetes1.24 package stopped to produce bit-identical results on 2 fresh builds.

```diff
--- strings RPMS.1/usr/bin/kubelet1.24
+++ strings RPMS.2/usr/bin/kubelet1.24
@@ -1,6 +1,6 @@
 /lib64/ld-linux-x86-64.so.2
-D/iB)
-YwVWBSSdZpPsrO9TM081/7CBb1fZOvPOFshzvo1cu/UKVjqzDsSfvmOjrwnpZv/UooIAWPloG5fXOlWD0oC
+rc{.
+tPfq4us5qtj6n5wGE3lh/7CBb1fZOvPOFshzvo1cu/UKVjqzDsSfvmOjrwnpZv/5Y18rAlngXbjuDxH_BT5
 _Z!#YH+
 o&43b
 =#jv
@@ -1456855,7 +1456855,7 @@
 /usr/lib64/go/1.18/src/net/tcpsockopt_unix.go
 /usr/lib64/go/1.18/src/net/udpsock_posix.go
 /usr/lib64/go/1.18/src/net/unixsock_posix.go
-/tmp/go-build1406627905/b050/_cgo_gotypes.go
+/tmp/go-build3558486601/b050/_cgo_gotypes.go
 /usr/lib64/go/1.18/src/net/cgo_resnew.go
 /usr/lib64/go/1.18/src/net/cgo_unix.go
 /usr/lib64/go/1.18/src/net/cgo_socknew.go
@@ -1457086,7 +1457086,7 @@
 vendor/k8s.io/klog/v2/internal/serialize/keyvalues.go
 /usr/lib64/go/1.18/src/os/user/lookup.go
 /usr/lib64/go/1.18/src/os/user/user.go
-/tmp/go-build1908184651/b122/_cgo_gotypes.go
+/tmp/go-build1410760062/b122/_cgo_gotypes.go
 /usr/lib64/go/1.18/src/os/user/cgo_lookup_unix.go
 vendor/k8s.io/klog/v2/exit.go
 vendor/k8s.io/klog/v2/klog.go
@@ -1460325,7 +1460325,7 @@
 vendor/github.com/google/cadvisor/utils/cpuload/cpuload.go
 vendor/github.com/docker/go-units/size.go
 vendor/github.com/docker/go-units/ulimit.go
-/tmp/go-build1382031104/b1017/_cgo_gotypes.go
+/tmp/go-build2774399619/b1017/_cgo_gotypes.go
 vendor/github.com/mindprince/gonvml/bindings.go
 vendor/github.com/google/cadvisor/accelerators/nvidia.go
 vendor/github.com/google/cadvisor/nvm/machine_no_libipmctl.go
```

The first chunk is likely just the hash of the remainder, so can be ignored.
The other chunks indicate that a random build temporary path is used for `_cgo_gotypes.go`

### What did you expect to happen?

Sources should be able to reproduce bit-identical binaries e.g. for SLSA level4

### How can we reproduce it (as minimally and precisely as possible)?

On Debian or openSUSE do
```bash
osc co openSUSE:Factory/kubernetes1.24 && cd $_
for i in 1 2 ; do
  osc build --clean --noservice --keep-pkg=RPMS.$i \
  --define='%source_date_epoch_from_changelog Y' \
  --define="%_buildhost reproducible" \
  --define='%clamp_mtime_to_source_date_epoch Y' \
  --define='%use_source_date_epoch_as_buildtime Y'
done
sha256sum RPMS.*/kubernetes1.24-kubelet-1*.rpm
```

### Anything else we need to know?

_No response_

### Kubernetes version

<details>

```console
$ kubectl version
```
WARNING: This version information is deprecated and will be replaced with the output from kubectl version --short.  Use --output=yaml|json to get the full version.
Client Version: version.Info{Major:"1", Minor:"24", GitVersion:"v1.24.0", GitCommit:"4ce5a8954017644c5420bae81d72b09b735c21f0", GitTreeState:"clean", BuildDate:"2022-05-05T00:00:00Z", GoVersion:"go1.18.3", Compiler:"gc", Platform:"linux/amd64"}
Kustomize Version: v4.5.4
```

</details>


### Cloud provider

<details>

</details>


### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
```
NAME="openSUSE Tumbleweed"
# VERSION="20220701"
ID="opensuse-tumbleweed"
ID_LIKE="opensuse suse"
VERSION_ID="20220701"
PRETTY_NAME="openSUSE Tumbleweed"
ANSI_COLOR="0;32"
CPE_NAME="cpe:/o:opensuse:tumbleweed:20220701"
BUG_REPORT_URL="https://bugs.opensuse.org"
HOME_URL="https://www.opensuse.org/"
DOCUMENTATION_URL="https://en.opensuse.org/Portal:Tumbleweed"
LOGO="distributor-logo-Tumbleweed"
```
$ uname -a
Linux bmwiedemann-factory 5.18.6-1-default #1 SMP PREEMPT_DYNAMIC Thu Jun 23 05:46:18 UTC 2022 (5aa0763) x86_64 x86_64 x86_64 GNU/Linux

```

</details>

### Install tools

<details>
go1.18
</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### PR DESCRIPTION
**Pull request by thockin**
Fixes #110928

/kind bug
```release-note
NONE
```

### CODE REVIEW

File: hack/lib/golang.sh [hunk 1]
```diff
@@ -865,26 +862,19 @@ kube::golang::build_binaries() {
...
-    fi
-
-    goldflags="all=$(kube::version::ldflags) ${GOLDFLAGS:-}"
-    if [[ "${DBG:-}" != 1 ]]; then
-        # Not debugging - disable symbols and DWARF.
+        gogcflags="${gogcflags} all=-N -l"
+    else
+        # Not debugging - disable symbols and DWARF, trim embedded paths
         goldflags="${goldflags} -s -w"
+        goflags+=("-trimpath")
```
Comment (dims) [inline hack/lib/golang.sh:877]: shouldn't there be the actual path too?

Comment (thockin) [reply hack/lib/golang.sh:877]: nope, `-trimpath` is a bool flag, it trims up to the module root.  Much simpler :)

Comment (dims) [reply hack/lib/golang.sh:877]: yay!

**Comment by dims** [comment]
/priority critical-urgent

Looks like we will be able to have reproducible builds (almost!) with this fix. will add milestone when the CI goes green.

/retest

File: hack/lib/golang.sh [hunk 2]
```diff
@@ -865,26 +862,19 @@ kube::golang::build_binaries() {
...
     # build_binaries_for_platform.
-    local goflags goldflags goasmflags gogcflags gotags
+    local goflags goldflags gogcflags gotags
 
-    # This is $(pwd) because we use run-in-gopath to build.  Once that is
-    # excised, this can become ${KUBE_ROOT}.
-    local trimroot # two lines to appease shellcheck SC2155
-    trimroot=$(pwd)
-
-    goasmflags="all=-trimpath=${trimroot}"
```
Comment (liggitt) [inline hack/lib/golang.sh:875]: was this just incorrect before? I thought we *wanted* to trim the root but leave relative source paths in place... those should still be reproducible, right?

Comment (thockin) [reply hack/lib/golang.sh:875]: From testing it looks like `-trimpath` does automatically what we were trying to do manually.

```
$ make kubectl DBG=1
+++ [1115 08:39:32] Building go targets for linux/amd64
    k8s.io/kubernetes/cmd/kubectl (static)
thockin-glaptop5 kubernetes f72e3d build_trimpath /$ strings _output/bin/kubectl | grep 'k8s\.io.*\.go'$ | head -10
/home/thockin/src/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/github.com/spf13/pflag/bool.go
/home/thockin/src/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/github.com/spf13/pflag/bool_slice.go
/home/thockin/src/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/github.com/spf13/pflag/bytes.go
/home/thockin/src/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/github.com/spf13/pflag/count.go
/home/thockin/src/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/github.com/spf13/pflag/duration.go
/home/thockin/src/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/github.com/spf13/pflag/duration_slice.go
/home/thockin/src/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/github.com/spf13/pflag/flag.go
/home/thockin/src/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/github.com/spf13/pflag/float32.go
/home/thockin/src/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/github.com/spf13/pflag/float32_slice.go
/home/thockin/src/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/github.com/spf13/pflag/float64.go

$ make kubectl
$ strings _output/bin/kubectl | grep 'k8s\.io.*\.go'$ | head -10
k8s.io/apimachinery@v0.0.0/pkg/util/sets/string.go
k8s.io/apimachinery@v0.0.0/pkg/util/sets/set.go
k8s.io/klog/v2@v2.110.1/internal/severity/severity.go
k8s.io/klog/v2@v2.110.1/internal/buffer/buffer.go
k8s.io/klog/v2@v2.110.1/internal/clock/clock.go
k8s.io/klog/v2@v2.110.1/internal/dbg/dbg.go
k8s.io/klog/v2@v2.110.1/internal/serialize/keyvalues.go
k8s.io/klog/v2@v2.110.1/internal/serialize/keyvalues_slog.go
k8s.io/klog/v2@v2.110.1/internal/sloghandler/sloghandler_slog.go
k8s.io/klog/v2@v2.110.1/exit.go
```

Note, it also changes paths into modules, but that seems OK?  If you need to debug (good old dlv!) then you need to map source files manually.  It probably chages the output of logging when the full caller is logged (rare).

Also note:

```
$ strings _output/bin/kubectl | grep thockin
```

No output on the non-DBG

Comment (thockin) [reply hack/lib/golang.sh:875]: More simply:

```
$ make kubectl
+++ [1115 08:45:53] Building go targets for linux/amd64
    k8s.io/kubernetes/cmd/kubectl (static)

$ strings _output/bin/kubectl | grep 'kubectl\.go'
k8s.io/kubernetes/cmd/kubectl/kubectl.go

$ make kubectl DBG=1
+++ [1115 08:46:16] Building go targets for linux/amd64
    k8s.io/kubernetes/cmd/kubectl (static)

$ strings _output/bin/kubectl | grep 'kubectl\.go'
/home/thockin/src/kubernetes/_output/local/go/src/k8s.io/kubernetes/cmd/kubectl/kubectl.go
```

**Comment by thockin** [comment]
The test-failure is verify-conformance-yaml throwing "Error looking for conformance data: open k8s.io/kubernetes/test/e2e/apimachinery/webhook.go: no such file or directory"

So changing it from a local filepath into a kube module name DOES have some impact.  Investigating

**Comment by thockin** [comment]
So, gingko uses runtime.Caller() which this changes.  Go team says "You can get the file path for a give package import path..."  which is not obvious from code.

**Comment by thockin** [comment]
I re-pushed with a hack to use DBG=1 for conformance verificaction.

**Comment by dims** [comment]
```
fmt.Println("TIM:", frame.File)
```

image

**Comment by thockin** [comment]
I have no idea what you mean.  I would _never_ leave that in the code.

**Comment by dims** [comment]
> I have no idea what you mean. I would _never_ leave that in the code.

just spotted that debug code :) i do this all the time as well!

**Comment by thockin** [comment]
All green

**Comment by dims** [comment]
still looks good!

/approve
/lgtm

**Comment by dims** [comment]
@thockin @liggitt do we want this for 1.29? can one of you please apply milestone if so.

**Comment by liggitt** [comment]
looks like it's existed in 1.24+, so I don't feel strongly if it goes in now or in 1.30 / 1.29.1

**Comment by thockin** [comment]
So if this isn't something we would backport, then we probably shouldn't push it to 29 ?

**Comment by thockin** [comment]
/retest

needs LGTM after rebase

**Comment by thockin** [comment]
still needs LGTM after rebase

**Comment by thockin** [comment]
Still needs LGTM

**Pull request merged**